src/mjolnir/bssbuilder.cc

```diff
@@ -0,0 +1,362 @@
+#include "mjolnir/bssbuilder.h"
+#include "mjolnir/graphtilebuilder.h"
+
+#include <boost/filesystem/operations.hpp>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <list>
+#include <mutex>
+#include <queue>
+#include <set>
+#include <thread>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/foreach.hpp>
+#include <boost/range/algorithm.hpp>
+
+#include "baldr/datetime.h"
+#include "baldr/graphreader.h"
+#include "baldr/graphtile.h"
+#include "baldr/tilehierarchy.h"
+#include "filesystem.h"
+#include "mjolnir/osmnode.h"
+#include "midgard/sequence.h"
+#include "midgard/distanceapproximator.h"
+#include "midgard/logging.h"
+#include "midgard/util.h"
+
+using namespace valhalla::midgard;
+using namespace valhalla::baldr;
+using namespace valhalla::baldr::DateTime;
+using namespace valhalla::mjolnir;
+
+namespace {
+
+struct OSMConnectionEdge {
+  GraphId startnode = {};
+  GraphId endnode = {};
+  float length = 0;
+  uint64_t wayid = -1;
+  std::vector<std::string> names = {};
+  std::list<PointLL> shape = {};
+  uint32_t speed = 0;
+  boost::optional<PointLL> bss_ll = boost::none;
+  OSMConnectionEdge(GraphId startnode,
+                    GraphId endnode,
+                    float length,
+                    uint64_t wayid,
+                    const std::vector<std::string>& names,
+                    const std::list<PointLL>& shape,
+                    uint32_t speed,
+                    boost::optional<PointLL> bss_ll = boost::none): startnode(startnode),
+                        endnode(endnode), length(length), wayid(wayid), names(names), shape(shape), speed(speed), bss_ll(bss_ll)  {}
+
+  // operator < for sorting
+  bool operator<(const OSMConnectionEdge& other) const {
+    if (startnode.tileid() == other.startnode.tileid()) {
+      return startnode.id() < other.startnode.id();
+    } else {
+      return startnode.tileid() < other.startnode.tileid();
+    }
+  }
+};
+
+
+template<typename T>
+struct Finally {
+  T t;
+  explicit Finally(T t): t(t) {};
+  Finally() = delete;
+  Finally(Finally&& f) = default;
+  Finally(const Finally&) = delete;
+  Finally& operator=(const Finally&) = delete;
+  Finally& operator=(Finally&&) = delete;
+  ~Finally(){ t(); };
+};
+
+template<typename T>
+Finally<T> make_finally(T t){
+  return Finally<T>{t};
+};
+
+struct BSSConnections {
+  // Connections that start with a way node that is already present in the tile
+  std::vector<OSMConnectionEdge> waynode_conn;
+  // Connections that start with a bss node that is to be added into the tile
+  std::vector<OSMConnectionEdge> bssnode_conn;
+};
+
+} // namespace
+
+namespace valhalla {
+namespace mjolnir {
+
+// Add bss to the graph
+void BssBuilder::Build(const boost::property_tree::ptree& pt) {
```

Comment:
it would be really nice if there were a way to make this optional. in transit_builder we can look at the transit tiles location and see that there is nothing there. perhaps we can just add a section to the mjolnir config that says whether or not to connect the bike share stations or not?

Reply:
Yes, absolutely!